app/controllers/session_controller.rb

```diff
@@ -271,6 +284,22 @@ def current
     end
   end
 
+  def second_factor
+    RateLimiter.new(nil, "login-hr-#{request.remote_ip}", SiteSetting.max_logins_per_ip_per_hour, 1.hour).performed!
+    RateLimiter.new(nil, "login-min-#{request.remote_ip}", SiteSetting.max_logins_per_ip_per_minute, 1.minute).performed!
+    if user = User.find_by_username_or_email(params[:login])
+      unless user.confirm_password?(params[:password])
+        invalid_credentials
```

Comment:
Minor but why not `return invalid_credentials`? One line shorter and reads better